File: prototype/rails/src/app/domain/listings/listing_slug.rb
module Domain
  module Listings
    # The storefront URL a listing is shared under. Titles repeat between
    # sellers, so a slug that is already taken counts up until it is free.
    module ListingSlug
      FALLBACK = "listing".freeze

      module_function

      def base(title)
        title.to_s.parameterize.presence || FALLBACK
      end

      def first_free(title, taken)
        candidate = base(title)
        return candidate unless taken.include?(candidate)

        suffix = 2
        suffix += 1 while taken.include?("#{candidate}-#{suffix}")

        "#{candidate}-#{suffix}"
      end
    end
  end
end

File: prototype/node/src/app/core/listings/listing-slug.ts
const FALLBACK = 'listing'

export function slugBase(title: string): string {
  const slug = title.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '')
  return slug.length === 0 ? FALLBACK : slug
}

// Titles repeat between sellers, so a slug that is already taken counts up
// until it finds one that is free.
export function firstFreeSlug(title: string, taken: readonly string[]): string {
  const candidate = slugBase(title)
  if (!taken.includes(candidate)) {
    return candidate
  }

  let suffix = 2
  while (taken.includes(`${candidate}-${suffix}`)) {
    suffix += 1
  }
  return `${candidate}-${suffix}`
}
